% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/plotSNPpeakPairs.R
\name{plotSNPpeakPairs}
\alias{plotSNPpeakPairs}
\title{plotSNPpeakPairs}
\usage{
plotSNPpeakPairs(genotypeFile, SNPID, geneName, IP_BAMs, INPUT_BAMs, size.IP,
  size.INPUT, geneModel, libraryType = "opposite", center = mean, GTF,
  ZoomIn = NULL, adjustExprLevel = TRUE)
}
\arguments{
\item{genotypeFile}{The path to vcf file where genotype is available.}

\item{SNPID}{rsID of the SNP to be ploted
#' @param geneName The name (as defined in gtf file) of the gene you want to plot}

\item{size.IP}{The size factor for IP libraries}

\item{size.INPUT}{The size factor for INPUT libraries}

\item{geneModel}{The gene model generated by gtfToGeneModel() function}

\item{libraryType}{"opposite" for mRNA stranded library, "same" for samll RNA library}

\item{GTF}{gtf annotation as GRanges object. Can be obtained by GTF <- rtracklayer::import("xxx.gtf",format = "gtf")}

\item{adjustExprLevel}{Logic parameter determining whether adjust coverage so that input are at "same" expression level.}

\item{IP_BAM}{The bam files for IP samples}

\item{INPUT_BAM}{The bam files for INPUT samples}
}
