#' @title plotSNPpeakPairs
#' @param genotypeFile The path to vcf file where genotype is available.
#' @param SNPID rsID of the SNP to be ploted
#' #' @param geneName The name (as defined in gtf file) of the gene you want to plot
#' @param IP_BAM The bam files for IP samples
#' @param INPUT_BAM The bam files for INPUT samples
#' @param size.IP The size factor for IP libraries
#' @param size.INPUT The size factor for INPUT libraries
#' @param geneModel The gene model generated by gtfToGeneModel() function
#' @param libraryType "opposite" for mRNA stranded library, "same" for samll RNA library
#' @param GTF gtf annotation as GRanges object. Can be obtained by GTF <- rtracklayer::import("xxx.gtf",format = "gtf")
#' @param adjustExprLevel Logic parameter determining whether adjust coverage so that input are at "same" expression level.
#' @export
plotSNPpeakPairs <- function(genotypeFile,SNPID,geneName,IP_BAMs, INPUT_BAMs, size.IP, size.INPUT, geneModel, libraryType = "opposite", center = mean ,GTF,ZoomIn=NULL, adjustExprLevel = TRUE){
  
  geno <- .getGenotype(genotypeFile,SNPID)
  
  dosage_geno <- .genotypeDosage(geno[-c(1:9)])
  
  snp_chr <- geno[1]
  snp_loc <- geno[2]
  snp_ref <- geno[4]
  snp_alt <- geno[5]
  
  X <- gsub("0",paste0(snp_ref,snp_ref),as.character(dosage_geno))
  X <- gsub("1",paste0(snp_ref,snp_alt),as.character(X))
  X <- gsub("2",paste0(snp_alt,snp_alt),as.character(X))
  
  X <- factor(X,levels=c(paste0(snp_ref,snp_ref),paste0(snp_ref,snp_alt),paste0(snp_alt,snp_alt))[c(paste0(snp_ref,snp_ref),paste0(snp_ref,snp_alt),paste0(snp_alt,snp_alt))%in%X])
  
  plotGeneCoverage(IP_BAMs, INPUT_BAMs, size.IP, size.INPUT,X, geneName, geneModel, libraryType, center  ,GTF,ZoomIn, adjustExprLevel, plotSNP = data.frame(loc = as.numeric(snp_loc), anno = paste0(snp_ref,"/",snp_alt) )  )+
    ggtitle(paste0("SNPID: ",SNPID,"    m6A peak on: ",geneName))+theme(plot.title = element_text(hjust = 0.5,size = 15,face = "bold"),legend.title =  element_text(hjust = 0.5,size = 13,face = "bold"),legend.text =  element_text(size = 12,face = "bold"))
  
}


.genotypeDosage <- function(x){
  split <- sapply(x,strsplit,"[|]")
  return( suppressWarnings( unlist(lapply(split,function(x){sum(as.integer(x),na.rm = T)})) ) )
}

.getGenotype <- function(vcf.gz, SNPID){
   tmp <- system2(command = "zcat", args = paste0(genotypeFile," |grep '",SNPID,"' |cat"),stdout = T)
   geno <- strsplit(tmp,split = "\t")[[1]]
   return(geno)
}